components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
+                    self.set_value_source_size_list(& mut source_set.source_size);
+                }
+                _ => ()
+            }
+
+            // Step 4.8
+            match element.get_attribute(&ns!(), &local_name!("type")) {
+                Some(x) => {
+                    match get_mime_type_str(&x.value()) {
+                        Some(_valid_mime) => (),
+                        None => continue // Unsupported mime type
+                    }
+                }
+                _ => continue // Unknown mime type
+            }
+
+            // Step 4.9
+            Self::normalise_source_densities(&mut source_set);
+
+            // Step 4.10
+            *self.source_set.borrow_mut() = source_set;
+            return ;
+        }
+    }
+
+    fn set_value_source_size_list(&self, source_size_list: &mut SourceSizeList) {
+        let document = document_from_node(self);
+        let device = document.device().unwrap();
+        let quirks_mode = document.quirks_mode();
+        let length = source_size_list.evaluate(&device, quirks_mode);
+        if !source_size_list.value.is_some() {
+            source_size_list.value = Some(Length::from_px(length.to_f32_px()));
         }
-        vec![src]
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#matches-the-environment
+    fn matches_environment(&self, media_query: String) -> bool {
+        let document = document_from_node(self);
+        let device = document.device().unwrap();
+        let quirks_mode = document.quirks_mode();
+        let document_url = &document.url();
+        let context = ParserContext::new(
+            Origin::Author,
+            document_url,
+            Some(CssRuleType::Style),
+            ParsingMode::all(),
+            quirks_mode,
+            None,
+        );
+        let mut parserInput = ParserInput::new(&media_query);
+        let mut parser = Parser::new(&mut parserInput);
+        let media_list = MediaList::parse(&context, &mut parser);
+        media_list.evaluate(&device, quirks_mode)
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#normalise-the-source-densities>
+    fn normalise_source_densities(source_set: &mut SourceSet) {
+        let length = &source_set.source_size.value.clone().unwrap();
+        let source_size_length_result = length.to_pixel_length(Some(Au::new(1 as i32)));
+        for imgsource in &mut source_set.image_sources {
+            if imgsource.descriptor.den.is_some() {
+                continue;
+            } else {
+                    if imgsource.descriptor.wid.is_some() {
+                        let source_size_length = match source_size_length_result {
+                            Ok(x) => x,
+                            _ => 1 as f32,
+                        };
+                        let wid = imgsource.descriptor.wid.unwrap();
+                        imgsource.descriptor.den = Some((wid / source_size_length as u32).into());
```

Comment:
Let's preserve our floating point calculations by casting `wid` to `f32` instead of casting `source_size_length` to u32. What is the `.into()` needed for?

Reply:
into is used for converting the result into f64 value